### PR TITLE
Object Storage: Support Alibaba cloud OSS provider

### PR DESCRIPTION
**Pull request by wujinhu**
## Changes
Alibaba cloud OSS(Object Storage Service) is widely used among China’s cloud users. This PR adds supports of OSS for thanos.

## Verification

- To run tests against this provider, you should set environment variables like below:
    export OSS_ENDPOINT='OSS endpoint to connect to'
    export OSS_ACCESS_KEY='Your access key'
    export OSS_SECRET_KEY='Your secret key'
- To use it in POC or production, you should add below to configuration file
**type**: OSS
**config**:
&emsp;**bucket**: "Your bucket name"
&emsp;**endpoint**: "OSS endpoint to connect to"
&emsp;**access_key**: "Your access key"
&emsp;**secret_key**: "Your secret key"

I'm from Alibaba Cloud and would love to become the maintainer for the OSS provider. :)

### CODE REVIEW

**Comment by wujinhu** [comment]
Looking into tests failed.

**Comment by wujinhu** [comment]
@bwplotka @jojohappy @yeya24 Could you please help review this PR, our users need this support, thanks.:)

**Comment by wujinhu** [comment]
=== RUN   TestObjStore_AcceptanceTest_e2e
--- PASS: TestObjStore_AcceptanceTest_e2e (2.05s)
=== RUN   TestObjStore_AcceptanceTest_e2e/inmem
    --- PASS: TestObjStore_AcceptanceTest_e2e/inmem (0.05s)
    foreach.go:49: THANOS_SKIP_GCS_TESTS envvar present. Skipping test against GCS.
    foreach.go:71: THANOS_SKIP_S3_AWS_TESTS envvar present. Skipping test against S3 AWS.
    foreach.go:87: THANOS_SKIP_AZURE_TESTS envvar present. Skipping test against Azure.
    foreach.go:103: THANOS_SKIP_SWIFT_TESTS envvar present. Skipping test against swift.
    oss.go:341: created temporary OSS bucket for OSS tests with name test-testobjstore-acceptancetest-e2e-1ee0ded3a2691b90
=== RUN   TestObjStore_AcceptanceTest_e2e/Alibaba_OSS
    --- PASS: TestObjStore_AcceptanceTest_e2e/Alibaba_OSS (0.86s)
    foreach.go:135: THANOS_SKIP_TENCENT_COS_TESTS envvar present. Skipping test against Tencent COS.
PASS

File: pkg/objstore/oss/oss.go [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/improbable-eng/thanos/pkg/runutil"
+
+	"github.com/aliyun/aliyun-oss-go-sdk/oss"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/improbable-eng/thanos/pkg/objstore"
```
Comment (yeya24) [inline pkg/objstore/oss/oss.go]: Use goimports to format this.  It should be above.

File: pkg/objstore/oss/oss.go [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+}
+
+func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	bucket, err := b.client.Bucket(b.name)
+	if err != nil {
+		return errors.Wrap(err, "upload oss object")
+	}
+
+	rc := ioutil.NopCloser(r)
+	err = bucket.PutObject(name, rc)
```
Comment (yeya24) [inline pkg/objstore/oss/oss.go]: Use if err := ; err != {} here.

File: pkg/objstore/oss/oss.go [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+// DirDelim is the delimiter used to model a directory structure in an object store bucket.
+const DirDelim = "/"
+
+// Config stores the configuration for OSS bucket.
+type Config struct {
+	Bucket     string     `yaml:"bucket"`
+	Endpoint   string     `yaml:"endpoint"`
+	AccessKey  string     `yaml:"access_key"`
+	SecretKey  string     `yaml:"secret_key"`
+	HTTPConfig HTTPConfig `yaml:"http_config"`
```
Comment (jojohappy) [inline pkg/objstore/oss/oss.go]: Just curious, what is the case let you setup the config of http client? why could we just use the default configuration?

File: pkg/objstore/oss/oss.go [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+	LogLevel   int        `yaml:"log_level"`
```
Comment (jojohappy) [inline pkg/objstore/oss/oss.go]: Would it print out oss log in thanos? IMO the error messages should be returned from oss package, and we can catch and throw it to the user. I don't know what do you concern for the oss logger. But IMO it is unnecessary, we should have a unified logger handler. What do you think?

File: pkg/objstore/oss/oss.go [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+		dir = strings.TrimSuffix(dir, DirDelim) + DirDelim
+	}
+
+	bucket, err := b.client.Bucket(b.name)
+	if err != nil {
+		return errors.Wrap(err, "iterate oss objects")
+	}
+
+	var marker = oss.Marker("")
+	for {
```
Comment (jojohappy) [inline pkg/objstore/oss/oss.go]: In a scenario, if the `ctx` is an instance of context with timeout, then it is timeout, would the `iterator` be stopped? So should we try to listen the channel of `ctx.Done()` to break the iterator? What do you think?

File: pkg/objstore/oss/oss.go [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+			}
+		}
+
+		for _, commonPrefixes := range lsRes.CommonPrefixes {
+			if err := f(commonPrefixes); err != nil {
+				return errors.Wrap(err, "iterate oss objects")
+			}
+		}
+
+		marker = oss.Marker(lsRes.NextMarker)
```
Comment (jojohappy) [inline pkg/objstore/oss/oss.go]: Should we move this next to the `if` statement?

File: pkg/objstore/oss/oss.go [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+		return nil, err
+	}
+
+	return r, nil
+}
+
+func (b *Bucket) IsObjNotFoundErr(err error) bool {
+	switch err.(type) {
+	case oss.ServiceError:
+		if err.(oss.ServiceError).StatusCode == 404 {
```
Comment (jojohappy) [inline pkg/objstore/oss/oss.go]:
```suggestion
		if err.(oss.ServiceError).StatusCode == http.StatusNotFound {
```

File: pkg/objstore/oss/oss.go [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+			return true
+		}
+		return false
+	default:
```
Comment (jojohappy) [inline pkg/objstore/oss/oss.go]: Remove the `default` case, just return false at the end of function.

**Comment by jojohappy** [comment]
@shaulboozhiao @wujinhu Thanks for your PR! Good job! But we could have only one implement of OSS, so could we make a discussion offline in slack?

**Comment by wujinhu** [comment]
Thanks @jojohappy @yeya24 for your suggestions，I will solve those issues. But first，I agree with you to make a discussion offline in slack about those whose two PR.：）

**Comment by shaulboozhiao** [comment]
@wujinhu  SO DISAPPOINTED WITH YOU ,  and SHAME ON YOU ~~~  Let me give you some advice in chinese "做事先做人！”

**Comment by wujinhu** [comment]
sorry，i did not know you have worked on this before i created this pr.If you want to continue work on this，i will close this pr

**Comment by wujinhu** [comment]
i have talked with @jojohappy offline，i will close this pr，please calm down！

**Pull request closed without merging**